Spark Java Framework Instrumentation

JorgenG opened the pull request:
This is a WIP of looking into sparkjava support.

tylerbenson commented on dd-java-agent/instrumentation/servlet-3/src/main/java/datadog/trace/instrumentation/servlet3/Filter3Instrumentation.java:
```diff
@@ -0,0 +1,183 @@
+import net.bytebuddy.asm.Advice;
+
+@AutoService(Instrumenter.class)
+public final class Filter3Instrumentation extends Instrumenter.Configurable {
```
While I don't love it, it does seem this is a best way to generically cover what you need.  The other option might be to instrument Jetty's [Handler.handle](https://github.com/eclipse/jetty.project/blob/master/jetty-server/src/main/java/org/eclipse/jetty/server/Handler.java#L76) interface.  Which do you think is better?

JorgenG replied:
Looking at the Jetty handler interface, I like that a lot more. I think that will fit better and have less edge cases like the Filter has. 

Should that go in a seperate package "jetty"?

tylerbenson commented on dd-java-agent/instrumentation/servlet-3/src/main/java/datadog/trace/instrumentation/servlet3/Filter3Instrumentation.java:
```diff
@@ -0,0 +1,183 @@
+          .buildSpan(SERVLET_OPERATION_NAME)
+          .asChildOf(extractedContext)
+          .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_SERVER)
+          .withTag(DDTags.SPAN_TYPE, DDSpanTypes.WEB_SERVLET)
```
Since this span could come from a potentially large number of classes, I'd like to have something to trace it back to the origin.  Please add a tag with the class name like so:
```
      .withTag("span.origin.type", statement.getClass().getName());
```

Would you mind also adding that to the other servlet-{2,3} instrumentation as well?
Thanks!

JorgenG replied:
What is `statement` here? Cannot find it, and nothing logically matches.

tylerbenson replied:
Sorry that was copy paste from the statement instrumentation. For jetty instrumentation it would be this.getClass().

tylerbenson commented on dd-java-agent/instrumentation/sparkjava/src/test/groovy/SparkJavaBasedTest.groovy:
```diff
@@ -0,0 +1,30 @@
+
+
+@Timeout(5)
+class SparkJavaBasedTest extends AgentTestRunner {
```
once you have the instrumentation working, I'd like to see this test suite include deeper verification and some edge cases.  Use [this](https://github.com/DataDog/dd-trace-java/blob/master/dd-java-agent/instrumentation/spring-web/src/test/groovy/test/SpringBootBasedTest.groovy) for inspiration.

tylerbenson commented:
Also, [it appears](https://github.com/perwendel/spark/commit/bdc1b1c2fdf0ae77ee106de5f5213946378a641e) spark dropped support for Java 7 in 2.0, so if we don't plan on supporting anything before then, you can exclude the tests from running on java 7 by doing [this](https://github.com/DataDog/dd-trace-java/blob/0812501aa3f831da861cc91719933b37a6e5a61a/dd-java-agent/instrumentation/okhttp-3/okhttp-3.gradle#L17-L26), except change the check to `JavaVersion.current().isJava8Compatible()`.

JorgenG commented:
Thanks for the good feedback. I do see that I have messed up a bit, so I want to clear some things up.

I had the impression that blindly adding instrumentation for `Filter3` would not be a good solution. So my original intention was to not have `Filter3Instrumentation` under servlet-3 (Although I managed to commit it). 

Does my view match your understanding? Or would having the `Filter3Instrumentation` be correct at that location, and then add spark specific instrumentation on spark-specific classes? 

Right now there are two "identical" files.

JorgenG commented on dd-java-agent/instrumentation/servlet-3/src/main/java/datadog/trace/instrumentation/servlet3/Filter3Instrumentation.java:
```diff
@@ -0,0 +1,183 @@
+        not(isInterface()).and(hasSuperType(named("javax.servlet.Filter"))),
+        classLoaderHasClasses("javax.servlet.AsyncEvent", "javax.servlet.AsyncListener"))
+      .transform(
+        new HelperInjector(
```
What are all these opentracing things used for?

JorgenG commented:
I have made some changes now.

- Removed `Filter3Instrumentation`
- Added `HandlerInstrumentation` for Jetty
- Added `RoutesInstrumentation` for enriching with match url from SparkJava

Once the overall structure is ok, I'll take another look at the tests etc.

JorgenG commented:
I have attempted to use google java format in the project, but the verification step still fails. Any tips?

tylerbenson commented:
There’s a gradle task you can run to format everything. I think it’s googleFormatJava or something like that.

tylerbenson commented on dd-java-agent/instrumentation/sparkjava/sparkjava.gradle:
```diff
@@ -0,0 +1,33 @@
+apply from: "${rootDir}/gradle/java.gradle"
+
+if (JavaVersion.current() != JavaVersion.VERSION_1_8) {
```
Change to isJava8Compatible()

JorgenG replied:
Fixed.

tylerbenson commented on dd-java-agent/instrumentation/sparkjava/src/main/java/datadog/trace/instrumentation/sparkjava/RoutesInstrumentation.java:
```diff
@@ -0,0 +1,54 @@
+    return agentBuilder
+        .type(
+            is(named("spark.route.Routes")),
+            classLoaderHasClasses("spark.embeddedserver.jetty.EmbeddedJettyServer"))
```
The main reason for adding this line is to limit the versions of the sparkjava library we try to instrument. This list should match the verifyPresent block in a versionScan in the gradle file.

JorgenG replied:
Added version scan for sparkjava as well.

tylerbenson commented on dd-java-agent/instrumentation/jetty-9/src/main/java/datadog/trace/instrumentation/jetty9/HandlerInstrumentation.java:
```diff
@@ -0,0 +1,169 @@
+  public HandlerInstrumentation() {
+    super("jetty", "jetty-9");
+  }
+
```
Standard practice. Since this is new instrumentation, can you override the default enabled method to return false? We will change to enabled by default in a future release. Same applies to the spark instrumentation.

tylerbenson commented on dd-java-agent/instrumentation/jetty-9/jetty-9.gradle:
```diff
@@ -0,0 +1,35 @@
+
+  testCompile project(':dd-java-agent:testing')
+  testCompile group: 'org.eclipse.jetty', name: 'jetty-server', version: '9.4.6.v20170531'
+  testCompile group: 'org.eclipse.jetty', name: 'jetty-servlet', version: '9.4.6.v20170531'
```
Any reason this instrumentation wouldn’t work with older versions of jetty? We usually name the instrumentation with the earliest version it can support.

JorgenG replied:
Added support for oldest 8.0.0 where signature is the same for Handler.

JorgenG commented on dd-java-agent/instrumentation/jetty-9/jetty-9.gradle:
```diff
@@ -0,0 +1,35 @@
+  group = "javax.servlet"
+  module = 'javax.servlet-api'
+  legacyModule = "servlet-api"
+  versions = "[3.0,)"
```
I think this jetty hook requires the Async things from servlet?

JorgenG commented on dd-java-agent/instrumentation/jetty-9/src/main/java/datadog/trace/instrumentation/jetty9/HandlerInstrumentation.java:
```diff
@@ -0,0 +1,169 @@
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static Scope startSpan(
+        @Advice.Argument(0) final String target, @Advice.Argument(2) final HttpServletRequest req) {
+
```
Added.

JorgenG commented:
I will work on some tests later this week.

tylerbenson commented on dd-java-agent/instrumentation/sparkjava-2.3/src/test/groovy/SparkJavaBasedTest.groovy:
```diff
@@ -0,0 +1,69 @@
+class SparkJavaBasedTest extends AgentTestRunner {
+
+  static {
+    TestSparkJavaApplication.initSpark()
```
I think I'd like to see this as a `setupSpec()` method with a corresponding `cleanupSpec()` method to shut it down. (Is it possible to shut down programmatically?)

JorgenG replied:
Fixed.

tylerbenson commented on dd-java-agent/instrumentation/sparkjava-2.3/src/test/groovy/SparkJavaBasedTest.groovy:
```diff
@@ -0,0 +1,69 @@
+    span.context().tags["http.status_code"] == 200
+    span.context().tags["thread.name"] != null
+    span.context().tags["thread.id"] != null
+    span.context().tags.size() == 8
```
I think this would look nicer if `span.context()` was declared as a variable and maybe even the tags.

JorgenG replied:
Partially fixed by extracting to a variable.

tylerbenson commented:
After a clean build, I think this is mostly ready.  I'm going to pass over to @realark for secondary review.

JorgenG commented:
My problem right now is that the basic test does not work. I think this is because my understanding of how to activate the instrumentation is wrong. 

Since the tests are failing, we can't verify if it works. Since `defaultEnabled()` returns  `false`, how will it be an active instrumentation in test setting?

JorgenG commented:
In the latest commit, we can observe that Spark is started in the tests, however no traces are registered, thus our tests don't work. How can we activate tracing for the default disabled instrumentation?

JorgenG commented:
Now I don't understand anything. Why are the tests succeeding on circle ci, but locally when i run `gradle test` it fails?

realark commented on dd-java-agent/instrumentation/sparkjava-2.3/sparkjava-2.3.gradle:
```diff
@@ -0,0 +1,43 @@
+
+dependencies {
+  sourceCompatibility = "1.8"
+  targetCompatibility = "1.8"
```
Looks like this instrumentation is only intended to apply to Java 8 and above, correct? If so, I think we have a problem. Attempting to load a 1.8 class on a 1.7 jvm, will cause an exception to be thrown when we initialize the instrumentation.

@JorgenG There's nothing for you to do about this (unless you can change source and target compatibility to 1.7?). @tylerbenson and I will have to figure out a solution. I'm sure we'll want to do this often, especially as more frameworks are compiled against newer jvms.

tylerbenson replied:
Maybe we could configure the tests to be Java 8 but leave the instrumentation at Java 7?

realark commented:
@JorgenG You can enable the instrumentation in the test using a static initializer to set the system property.

Kafka example: https://github.com/DataDog/dd-trace-java/blob/master/dd-java-agent/instrumentation/kafka-clients-0.11/src/test/groovy/KafkaClientTest.groovy#L23-L25

I think you'll want something like this:

```
  static {
    System.setProperty("dd.integration.sparkjava.enabled", "true")
  }
```

JorgenG commented:
Attempted to add the enabling of instrumentation in tests, not able to get it working. I am leaving for two weeks vacation on thursday, and will be back at beginning of April. Will not be able to do any contributions during this time, so if you are able to provide support on getting the tests running that would be nice.

realark commented:
@JorgenG Thanks for your patience while we work on unblocking you. We're still working on a solution for Java 8 instrumentation, but we'll let you know as soon as we can get this merged.

realark commented on dd-java-agent-ittests/src/test/groovy/datadog/trace/agent/integration/classloading/ShadowPackageRenamingTest.groovy:
```diff
@@ -9,7 +9,7 @@ import spock.lang.Timeout
 import java.lang.reflect.Field
 
-@Timeout(1)
+@Timeout(10)
```
:100:

realark commented:
Looks good!

tylerbenson commented:
@JorgenG Thanks for your contribution.  Glad we could finally get this merged.  Sorry for the delay.

JorgenG commented:
Thanks for all the help, and happy to see it getting in. 👍
